bgpd/bgp_vty.c

```diff
@@ -3689,43 +3689,96 @@ DEFUN (no_bgp_neighbor_graceful_restart_disable,
 	return bgp_vty_return(vty, ret);
 }
 
-DEFPY (neighbor_graceful_shutdown,
-       neighbor_graceful_shutdown_cmd,
-       "[no$no] neighbor <A.B.C.D|X:X::X:X|WORD>$neighbor graceful-shutdown",
-       NO_STR
-       NEIGHBOR_STR
-       NEIGHBOR_ADDR_STR2
-       "Graceful shutdown\n")
+/*
+ * Function to announce route to peer
+ */
+static void bgp_peer_announce_routes(struct peer *peer)
 {
 	afi_t afi;
 	safi_t safi;
-	struct peer *peer;
-	int ret;
+	struct peer_af *paf = NULL;
+
+	FOREACH_AFI_SAFI (afi, safi) {
+		if (!peer->afc[afi][safi])
+			continue;
+		paf = peer_af_find(peer, afi, safi);
+		if (paf) {
+			update_group_adjust_peer(paf);
+			bgp_announce_route(peer, afi, safi, false);
+		}
+	}
+}
+
+/*
+ * Function to perform a soft reset of BGP neighborship on a peer or peer group
+ */
+static int bgp_peer_soft_reset(struct vty *vty, const char *peer_str, enum clear_sort sort)
+{
+	struct peer *peer = NULL;
+	struct listnode *node, *nnode;
+	struct peer_group *group;
 
 	VTY_DECLVAR_CONTEXT(bgp, bgp);
 
-	peer = peer_and_group_lookup_vty(vty, neighbor);
+	peer = peer_and_group_lookup_vty(vty, peer_str);
+	bgp_clear(vty, bgp, AFI_UNSPEC, SAFI_UNSPEC, sort, BGP_CLEAR_SOFT_IN, peer_str);
+
+	if (sort == clear_group) {
+		group = peer->group;
+		for (ALL_LIST_ELEMENTS(group->peer, node, nnode, peer))
+			bgp_peer_announce_routes(peer);
+	} else {
+		bgp_peer_announce_routes(peer);
+	}
+	return CMD_SUCCESS;
+}
+
+/*
+ * Function to enable or disable graceful shutdown configuration
+ * on a peer or peer group
+ */
+static int bgp_peer_gshut_set_vty(struct vty *vty, const char *peer_str, bool set)
```

Comment:
TBH, the keyword `set` here is confusing, it also does `unset`. Let's drop this function at all and reuse peer_flag_set_vty/peer_flag_unset_vty directly.